site/src/hooks/useClipboard.ts

```diff
@@ -1,44 +1,178 @@
+    setClipboardText((current) => (result.success ? result.value : current));
+  });
+
+  useEffect(() => {
+    // Focus event handles case where user navigates to a different tab, copies
+    // new text, and then comes back to Coder
+    window.addEventListener("focus", syncClipboardToState);
+    window.addEventListener("copy", syncClipboardToState);
+    void syncClipboardToState();
+
+    return () => {
+      window.removeEventListener("focus", syncClipboardToState);
+      window.removeEventListener("copy", syncClipboardToState);
+    };
+  }, [syncClipboardToState]);
+
+  const copyToClipboard = useCallback(async () => {
+    const result = await writeToClipboard(textToCopy);
+    if (result.success) {
+      void syncClipboardToState();
+      return;
     }
-  };
+  }, [syncClipboardToState, textToCopy]);
 
   return {
-    isCopied,
-    copy,
+    copyToClipboard,
+    isCopied: textToCopy === clipboardText,
   };
 };
+
+type VoidResult = Readonly<
+  { success: true; error: null } | { success: false; error: Error }
+>;
+
+type ResultWithData<T = unknown> = Readonly<
+  | { success: true; value: T; error: null }
+  | { success: false; value: null; error: Error }
+>;
+
+type Result<T = unknown> = void extends T ? VoidResult : ResultWithData<T>;
+
+async function readFromClipboard(): Promise<Result<string>> {
+  // This is mainly here for the sake of being exhaustive, but the main thing it
+  // helps with is suppressing error messages when Vite does HMR refreshes in
+  // dev mode
+  if (!document.hasFocus()) {
+    return {
+      success: false,
+      value: null,
+      error: new Error(
+        "Security error - clipboard read queued while tab was not active",
+      ),
+    };
+  }
+
+  try {
+    // navigator.clipboard is a newer API. It should be defined in most browsers
+    // nowadays, but there's a fallback if not
+    if (typeof window?.navigator?.clipboard?.readText === "function") {
+      return {
+        success: true,
+        value: await window.navigator.clipboard.readText(),
+        error: null,
+      };
+    }
+
+    const { isExecSupported, value } = simulateClipboard("read");
+    if (!isExecSupported) {
+      throw new Error(
+        "document.execCommand has been removed for the user's browser, but they do not have access to newer API",
+      );
+    }
+
+    return {
+      success: true,
+      value: value,
+      error: null,
+    };
+  } catch (err) {
+    // Only expected error not covered by function logic is the user not
+    // granting the webpage permission to access the clipboard
+    const flattenedError =
+      err instanceof Error
+        ? err
+        : new Error("Unknown error thrown while reading");
+
+    return {
+      success: false,
+      value: null,
+      error: flattenedError,
+    };
+  }
+}
+
+// Comments for this function mirror the ones for readFromClipboard
+async function writeToClipboard(textToCopy: string): Promise<Result<void>> {
+  if (!document.hasFocus()) {
```

Comment:
Hmm... why is this necessary?

Thinking about the most common use case, like the user clicking the button and the value being copied, I don't see how this could happen if the document is not in focus 🤔

Reply:
I had a comment in `readToClipboard`, but it's mainly to suppress errors when trying to do stuff in development mode

Reply:
Do you think is there a way to improve readability without adding a comment mentioning another comment? Thinking on a named variable or function like `isDevMode`?

Reply:
Yeah, now that you mention it, I can use an assertion function. I'll get that swapped in